### PR TITLE
[distributed launcher] auto find free port

### PR DESCRIPTION
**Pull request by stas00**
This PR tries to fix an issue of "Address is in use" error which happens when `python -m torch.distributed.launch` is used concurrently with default arguments - and no explicit port is passed. This happens e.g. on CI if there is some run-away process/test and then other processes can't launch the distributed anymore.

Because it's not enough to check the port being free, since there could be a race condition, in the proposed PR I added a randomization of the port which should reduce such possibility to very small chance. The random range can be made bigger to minimize this chance even further.

The only thing I'm not sure about is that now the default port is not necessarily 29500, which may break some systems if they hardcoded this number.

So if the proposal as I coded doesn't work, perhaps it'd work to add an explicit `--find_free_port` instead which would then try to pick a random port and try to see if it's open. Of course, this still doesn't prevent the possible race condition. So neither this solution is perfect.

As I'm writing this out I think I convinced myself, that perhaps leaving it to the user to sort out the explicit port as it is now is still the best course of action as they have full control over their environment. 

I just thought that perhaps things could be made a bit do-what-i-mean wrt distributed.

I will close this PR in a few days unless someone sees a value in it.

Thank you for reading.

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
@stas00 your proposal here works better than a fixed default port. The codes overall look good to me. But I would like to invite some network experts in our team to further review the diff later on. thanks for contributing this!

File: torch/distributed/launch.py [hunk 1]
```diff
@@ -174,10 +174,10 @@ def parse_args():
                              "the IP address or the hostname of node 0, for "
                              "single node multi-proc training, the "
                              "--master_addr can simply be 127.0.0.1")
-    parser.add_argument("--master_port", default=29500, type=int,
+    parser.add_argument("--master_port", default=0, type=int,
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:177]: Can we keep the default the same to avoid breaking any existing applications? We can extend the docs to mention that if you pass in 0, we will find an appropriate free port for you.

This way users that would like to take advantage of this can just pass in 0 for master port.

File: torch/distributed/launch.py [hunk 2]
```diff
@@ -212,6 +212,33 @@ def parse_args():
     parser.add_argument('training_script_args', nargs=REMAINDER)
     return parser.parse_args()
 
+from random import randint
+import socket
+def is_port_in_use(master_address, master_port):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        return s.connect_ex((master_address, master_port)) == 0
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:219]: Shouldn't we try `bind()` here instead of connect? Wouldn't the connect always fail since nothing is listening on that port yet?

File: torch/distributed/launch.py [hunk 3]
```diff
@@ -221,7 +248,7 @@ def main():
     # set PyTorch distributed related environmental variables
     current_env = os.environ.copy()
     current_env["MASTER_ADDR"] = args.master_addr
-    current_env["MASTER_PORT"] = str(args.master_port)
+    current_env["MASTER_PORT"] = str(get_port(args.master_addr, args.master_port))
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:251]: I don't think we need `get_port`, just setting "MASTER_PORT" to `args.master_port` should suffice. When "MASTER_PORT" is 0, it would pass this value all the way down to TCPStore: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/Utils.cpp#L80, which would listen on port 0 which would automatically pick a free port.

This also gets rid of the race as well. I'm wondering if we need any code change at all here since `--master_port 0` should work appropriately? We probably just need to update the docs to reflect this?

File: torch/distributed/launch.py (hunk 1 above)
Comment (stas00) [reply torch/distributed/launch.py:177]: That sounds good.

File: torch/distributed/launch.py (hunk 2 above)
Comment (stas00) [reply torch/distributed/launch.py:219]: Did you mean to say "always succeed"

We are just testing if the port is free here, and unfortunately I can't see how to avoid the race condition here.

Unless I misunderstood your comment, @pritamdamania87

File: torch/distributed/launch.py (hunk 3 above)
Comment (stas00) [reply torch/distributed/launch.py:251]: Oh, I had no idea it was that simple - then yes, by all means, this is a much more superior solution. 

I wonder why it wasn't used in first place, in used a hardcoded 29500 port instead.

Comment (stas00) [reply torch/distributed/launch.py:251]: probably should also document that 29500 is the default port - I don't think it says that anywhere in the public docs.

Comment (pritamdamania87) [reply torch/distributed/launch.py:251]: Actually there is a problem with passing in 0 as master port that I missed while writing the original comment. This would only work for a single process. Passing in a port of 0 means that every process will probably end up choosing a different port and that would cause rendezvous to fail since all processes need to agree on a single master port and address that they need to connect to.

I couldn't think of a nice way to enable dynamic ports here since the static master port and address is essentially how we ensure all processes can start communicating in the first place.

Comment (stas00) [reply torch/distributed/launch.py:251]: Oh bummer, I thought it was some magical way, but as you explain it - it's not.

To solve the race solution there needs to be some sort of port server that can allocate and hold the port for the client to connect later.

The impetus for this PR was a need for more than one test in the test suite being able to launch ddp.

Comment (pritamdamania87) [reply torch/distributed/launch.py:251]: > The impetus for this PR was a need for more than one test in the test suite being able to launch ddp.

I see, in that case I guess the test_suite itself should probably perform some logic similar to `get_port` in this PR and pass an appropriate MASTER_PORT to the launcher.

Comment (stas00) [reply torch/distributed/launch.py:251]: Totally, but I thought why re-invent the wheel thousands of times for each user when the framework could provide this solution.

That's why my 2nd proposal in OP, is to provide a flag which will randomize and check the port. this won't avoid the race condition though so I wasn't sure this is good enough still. And having an unreliable test suite isn't good. So the test suite needs to solve it on the client side with some internal counter then.

So I'm totally fine if this PR is put to rest as an experiment.
